### ISSUE TITLE
Failed to string_visible_contains and string_visible_exact under sdk level 17

### ISSUE THREAD

**Issue by KazuCocoa** [appium/ruby_lib#316]
`new UiSelector().resourceId()` is only provided over sdk level 18.
So, current `string_visible_contains` and `string_visible_exact` are failed if we use android under sdk level 17.

I would like to fix it and send PR to this repository, but I couldn't find method how to identify the target device's sdk version.I suppose I can use `driver_attributes[:caps][:platformVersion]`, but i would like to use more suitable method if you have.

Do you have any methods to identify the target sdk version ?

- Just example
```ruby
# To identify target sdk is over API Level 18 or not. 4.2 is API Level 17
def android_sdk_over_18?
  driver_attributes[:caps][:platformVersion].to_f > 4.2
end

# return resoirceId as additional UiSelector if target sdk is over API Level 18.
def string_visible_exact class_name, value
  value = %Q("#{value}")

  if class_name == '*'
    elements = "new UiSelector().description(#{value});" +
      "new UiSelector().text(#{value});"
    elements += "new UiSelector().resourceId(#{value});" if android_sdk_over_18?
    return elements
  end

  class_name = %Q("#{class_name}")

  elements = "new UiSelector().className(#{class_name}).text(#{value});" +
    "new UiSelector().className(#{class_name}).description(#{value});"
  elements += "new UiSelector().className(#{class_name}).resourceId(#{value});" if android_sdk_over_18?
  elements
end
```

I thinks `scroll_to` and `scroll_to_exact` also support resourceId as well.

**Comment by bootstraponline** [comment]
> Do you have any methods to identify the target sdk version ?

Yes, in the appium server bootstrap we can [tell with certainty what the API level is](https://github.com/appium/appium/blob/37afd9842c37598f2180172ba183ff1413ddf4f9/lib/devices/android/bootstrap/src/io/appium/android/bootstrap/utils/API.java#L8) because the code executes on device. From the Ruby bindings perspective, we're not actually on the device and it's unclear what API level is being used.

This is considered a multi-selection scenario where more than one [UiSelector is provided](https://github.com/appium/appium/blob/37afd9842c37598f2180172ba183ff1413ddf4f9/lib/devices/android/bootstrap/src/io/appium/android/bootstrap/utils/UiSelectorParser.java). The appium server should simply ignore the resourceId selector on devices that are <= 18 when we've given other selectors that are valid for the current API level. This approach would also fix the problem in all the bindings instead of just for Ruby.

**Comment by bootstraponline** [comment]
There are also some related issues in the server that cause UiSelector parsing to fail:
- https://github.com/appium/appium/issues/4311
- https://github.com/appium/appium/issues/4482

**Comment by bootstraponline** [comment]
This is a dupe of #232

**Comment by bootstraponline** [comment]
I think the approach I mentioned in #232 (client side detection using a regex) is going to be the easiest. The proper server fix will likely take longer to be implemented.

**Comment by bootstraponline** [comment]
@xsmaster @KazuCocoa This [patch](https://github.com/appium/ruby_lib/commit/23421e6a99d6f551e56e0de7751e6660ef2f53db) should help. It changes the default behavior of always searching for resourceId (which crashes on API < 18), to only search when the resourceId is valid.

**Comment by KazuCocoa** [comment]
> This is considered a multi-selection scenario where more than one UiSelector is provided. The appium server should simply ignore the resourceId selector on devices that are <= 18 when we've given other selectors that are valid for the current API level. This approach would also fix the problem in all the bindings instead of just for Ruby.

Oh, the solution is best I think. :+1: 


I saw your patch. It looks good. Thanks for your quick support.
I'll use current master branch until ruby_lib will release.

**Comment by xsmaster** [comment]
Thanks @bootstraponline !!!!

**Comment by bootstraponline** [comment]
I'll make a new ruby_lib release as soon as the appium server is released (I think it'll be 1.3.8). The current master branch has an iOS fix that requires appium master, otherwise I'd make a release now.

I opened some issues upstream on the server where the devs are considering some ways to make this a better experience. Thanks for reporting the bug!

https://github.com/appium/appium/issues/4858

**Comment by KazuCocoa** [comment]
:+1:
